src/Control/Monad/Aff.js

```diff
@@ -293,3 +293,70 @@ exports._liftEff = function (nonCanceler, e) {
     return nonCanceler;
   };
 }
+
+exports._tailRecM = function (isLeft, f, a) {
+  return function(success, error) {
+    return function go(acc) {
+      var result, status, canceler;
+
+      // Observes synchronous effects using a flag.
+      //
+      //   status = 0 (unresolved status)
+      //   status = 1 (synchronous effect)
+      //   status = 2 (asynchronous effect)
+      //
+      // The inner Aff is run, and if the status is still 0, we have observed
+      // a synchronous effect. The result is stashed for later synchronous
+      // processing. In the event that the status is 2, we have an asynchronous
+      // effect, and we can continue as normal, knowing we are on a different
+      // stack.
+
+      while (1) {
+        status = 0;
+        canceler = f(acc)(function(v) {
+          if (status === 0) {
+            result = v;
+            status = 1;
+          } else {
+            try {
+              if (isLeft(v)) {
+                go(v.value0);
```

Comment:
The recursion is in the "asynchronous" branch. It only is triggered when we have observed an asynchronous effect, and we are thus on a new stack.